kernel/src/fs/file_table.rs

```diff
@@ -162,6 +162,7 @@ impl FileTable {
             removed_entry.notify_fd_events(&events);
             closed_files.push(removed_entry.file.clone());
             if let Some(inode_file) = removed_entry.file.downcast_ref::<InodeHandle>() {
+                // FIXME: Operation below should not hold any mutex if `self` is protected by a spinlock externally
```

Comment:
So `release_range_locks` does use `Mutex` internally?

I think the bigger problem is that `Subject` contains `Mutex` internally. And both `FileTable` and `FileTableEntry` contain `Subject`. Even if `Subject` does not use `Mutex`, one can call `Subject::notify_observers` to trigger registered callbacks by `Observer`, which may use `Mutex` or other sleeping logic.

Reply:
I found

1. The `Subject` can be revised to fit the bill. Firstly the `observers` can be spin-locked, most operation is rather light-weight. The only concern is `notify_observers()`, and we can narrow down the lock scope, collect the active observers and do `observer.on_events(events)` after we unlock. (Done)

2. The `release_range_locks()` in `FileTable::close_file()` may hold mutex only when `fcntl` is used (the inner inode contains the extension of `RangeLockList`). This case can also be refactored as moving the mutex part out of the spin-locked scope, but requies more refactoring work then `Subject`, plus the usage in `fcntl` is rare, so I leave a `FIXME` here.